File: app/src/main/java/edu/uark/cartapp/models/api/enums/ProductApiMethod.java
package edu.uark.cartapp.models.api.enums;

import java.util.HashMap;
import java.util.Map;

import edu.uark.cartapp.models.api.interfaces.PathElementInterface;

/* ==== APP ProductApiMethod.java ====*/
public enum ProductApiMethod implements PathElementInterface {
	NONE(""),
	PRODUCT("product"),
	BY_LOOKUP_CODE("byLookupCode"),
	PRODUCTS("products");

	private String value;
	private static Map<String, ProductApiMethod> valueMap = null;

	ProductApiMethod(String value) { this.value = value; }

	public static ProductApiMethod map(String key) {
		if (valueMap == null) {
			valueMap = new HashMap<>();

			for (ProductApiMethod value : ProductApiMethod.values()) {
				valueMap.put(value.getPathValue(), value);
			}
		}
		return (valueMap.containsKey(key) ? valueMap.get(key) : ProductApiMethod.NONE);
	}

	@Override
	public String getPathValue() { return value; }
}

File: app/src/main/java/edu/uark/cartapp/models/api/enums/ProductApiMethod.kt
package edu.uark.cartapp.models.api.enums

import java.util.HashMap

import edu.uark.cartapp.models.api.interfaces.PathElementInterface

/* ==== APP ProductApiMethod.java ====*/
enum class ProductApiMethod private constructor(private val value: String) : PathElementInterface {
	NONE(""),
	PRODUCT("product"),
	BY_LOOKUP_CODE("byLookupCode"),
	PRODUCTS("products");

	override fun getPathValue(): String {
		return value
	}

	companion object {
		private var valueMap: MutableMap<String, ProductApiMethod>? = null

		fun map(key: String): ProductApiMethod? {
			if (valueMap == null) {
				valueMap = HashMap()

				for (value in ProductApiMethod.values()) {
					valueMap!![value.pathValue] = value
				}
			}
			return if (valueMap!!.containsKey(key)) valueMap!![key] else ProductApiMethod.NONE
		}
	}
}